circuits/cpp/src/aztec3/msgpack/schema_impl.hpp

```diff
@@ -0,0 +1,190 @@
+ * @param object The object in question.
+ */
+template <HasMsgPack T> inline void schema_pack(SchemaPacker& packer, T const& object)
+{
+    std::string type = schema_name(object);
+    if (packer.set_emitted(type)) {
+        packer.pack(type);
+        return; // already emitted
+    }
+    msgpack::check_msgpack_usage(object);
+    // Encode as map
+    const_cast<T&>(object).msgpack([&](auto&... args) {
+        size_t kv_size = sizeof...(args);
+        packer.pack_map(uint32_t(1 + kv_size / 2));
+        packer.pack("__typename");
+        packer.pack(type);
+        _schema_pack_map_content(packer, args...);
+    });
+}
+
+// Alias verification_key as verification_key_data
+inline void schema_pack(SchemaPacker& packer, proof_system::plonk::verification_key const&)
+{
+    schema_pack(packer, proof_system::plonk::verification_key_data{});
+}
+
+// Recurse over any templated containers
+// Outputs e.g. ['vector', ['sub-type']]
+template <typename... Args> inline void _schema_pack(SchemaPacker& packer, const std::string& schema_name)
+{
+    packer.pack_array(2);
+    packer.pack(schema_name);
+    packer.pack_array(sizeof...(Args));
+    (schema_pack(packer, Args{}), ...); /* pack schemas of all template Args */
+}
+template <typename... Args> inline void schema_pack(SchemaPacker& packer, std::tuple<Args...> const&)
+{
+    _schema_pack<Args...>(packer, "tuple");
+}
+template <typename K, typename V> inline void schema_pack(SchemaPacker& packer, std::map<K, V> const&)
+{
+    _schema_pack<K, V>(packer, "map");
+}
+template <typename T> inline void schema_pack(SchemaPacker& packer, std::optional<T> const&)
+{
+    _schema_pack<T>(packer, "optional");
+}
+template <typename T> inline void schema_pack(SchemaPacker& packer, std::vector<T> const&)
+{
+    _schema_pack<T>(packer, "vector");
+}
+template <typename... Args> inline void schema_pack(SchemaPacker& packer, std::variant<Args...> const&)
+{
+    _schema_pack<Args...>(packer, "variant");
+}
+template <typename T> inline void schema_pack(SchemaPacker& packer, std::shared_ptr<T> const&)
+{
+    _schema_pack<T>(packer, "shared_ptr");
+}
+
+// Outputs e.g. ['array', ['array-type', 'N']]
+template <typename T, std::size_t N> inline void schema_pack(SchemaPacker& packer, std::array<T, N> const&)
+{
+    packer.pack_array(2);
+    packer.pack("array");
+    packer.pack_array(2); /* param list format for consistency*/
+    schema_pack(packer, T{});
+    packer.pack(N);
+}
+
+/**
+ * @brief Print's an object's derived msgpack schema as a string.
+ *
+ * @param obj The object to print schema of.
+ * @return std::string The schema as a string.
+ */
+std::string schema_to_string(auto obj)
+{
+    msgpack::sbuffer output;
+    SchemaPacker printer{ output };
+    schema_pack(printer, obj);
+    msgpack::object_handle oh = msgpack::unpack(output.data(), output.size());
+    std::stringstream pretty_output;
+    pretty_output << oh.get() << std::endl;
+    return pretty_output.str();
```

Comment:
This looks familiar.... Is there something almost identical elsewhere?

Reply:
It's similar. I can make a to_json utility or something